Replacing AdjointMPI with Medi

talbring opened the pull request:
This pull request replaces AdjointMPI with our own open-source library called Medipack (https://github.com/SciCompKL/MeDiPack). It covers most of the MPI standard (especially non-blocking communication, custom data types and operators), which makes it more flexible for future extensions of the MPI communication in SU2.

Furthermore this updates Codi to the newest version and provides support for different taping strategies.

Eventually we want to get rid of the #ifdef's for MPI, so I also started with replacing almost all MPI calls with calls to the SU2_MPI wrapper class. 

I also removed the complex step datatype since I cannot make sure that it still works (and it makes no sense to maintain a forward mode and complex step ...)

economon commented on Common/src/config_structure.cpp:
```diff
@@ -153,9 +156,9 @@ unsigned short CConfig::GetnZone(string val_mesh_filename, unsigned short val_fo
-        MPI_Barrier(MPI_COMM_WORLD);
-        MPI_Abort(MPI_COMM_WORLD,1);
-        MPI_Finalize();
+        SU2_MPI::Barrier(MPI_COMM_WORLD);
```
Similar to my comment about rank/size, I would really like to see a cleaner, standard error handling. Doesn't have to be anything fancy, just a single routine, accessible everywhere (maybe in the MPI wrapper) that does something like the following:

SU2_MPI::Error("This is a string describing the error.", "CClassName::MethodName")

and inside will be the cout to dump the string and then the ifdef's for the correct exit in serial or parallel.

tobadavid replied:
Hi,

Following the same ideas, would it be possible to have a set/get mechanism for the communicator itself ? The CDriver takes a communicator as argument, but currently does nothing with it. The first motivation was to enable SU2 to be initialized by the Py wrapper with a specific communicator (e.g. a subset of MPI_COMM_WORLD). At the beginning the idea was to Set/Get this communicator from the CConfig class, but then we figured out that we were not able to access it everywhere in the code...if I remember well.

David

talbring replied:
I added a error function like you suggested @economon. Have a look at config_structure.cpp, I replaced already all of the "exit(...)" with that routine there.

@tobadavid, yep thats also something that should be possible.  I'll add a preliminary implementation soon.

talbring commented on .travis.yml:
```diff
@@ -20,7 +20,7 @@ branches:
 env:
     global:
-        CXXFLAGS="-O3 -Wall -Wextra -Wno-unused-parameter -Wno-empty-body -Wno-format-security" 
+        CXXFLAGS="-O3 -Wall -Wextra -Wno-unused-parameter -Wno-empty-body -Wno-format-security -DMEDI_NO_OPTIONAL_CONST" 
```
The problem is that depending on the actual MPI implementation (so for example MPICH, openmpi etc.) send buffer arguments might be declared as const or not. Unfortunately, there is no (easy) way to get that information automatically. It might be even different between versions of openmpi (on travis it is v1.6.5 and its necessary, locally I have v3.x, and its not)... so for now it is trial and error whether this is necessary or not. Maybe I can find that information somewhere, then I will provide it in the wiki.

talbring replied:
We found a solution for that. At least for OpenMPI and MPICH2. Still, there is no guarantee that it works for all MPI implementations (especially IntelMPI and PlatformMPI)... However, that should already cover the majority of setups.

talbring commented on Common/include/mpi_structure.hpp:
```diff
@@ -180,101 +230,112 @@ class CAuxMPIWrapper : public CMPIWrapper{
 
-  static void Allgather(void *sendbuf, int sendcnt, MPI_Datatype sendtype,
-                        void *recvbuf, int recvcnt, MPI_Datatype recvtype, MPI_Comm comm);
+  static void Allgather(void *sendbuf, int sendcnt, Datatype sendtype,
```
Yep, all of this collective communications should work now. They just have to be added to the wrapper.

vdweide replied:
In my opinion MPI_Allgatherv is rather essential to have, because the current workaround can be very memory consuming, especially when the number of ranks increases. There are other MPI routines that are used in the DG branches that are (were) not supported, but these are not used with su2double. So it would be nice to have them supported for consistency, but it does not have the highest priority. These are
- MPI_Reduce_scatter
- MPI_Probe
- MPI_Get_count
- MPI_Alltoall
- MPI_Testall

And then there are still the calls to MPI_Buffer_attach and MPI_Buffer_detach, which I have never understood.

talbring replied:
Actually I can remove the MPI_Buffer_attach/detach. I was using it before for my naive implementation of the forward mode MPI wrapper.

I will add the routines your are referring to.

talbring commented on Common/src/config_structure.cpp:
```diff
@@ -33,10 +33,13 @@
   
 #ifdef HAVE_MPI
-  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  SU2_MPI::Comm_rank(MPI_COMM_WORLD, &rank);
```
I already added a default wrapper implementation in this branch which simply copies the data for most of the mpi calls. So in theory we can already remove all the #ifdefs. However, like you said, I am not quite sure if that works for all parts, especially in geometry_structure.cpp with all the non-blocking communication. That's the reason why I haven't done it yet.

About the second aspect you are mentioning: I also thought about that. Another possibility would be to put this stuff in the wrapper. So that you can simply do SU2_MPI::Get_rank()/SU2_MPI::Get_size() or something.

economon commented:
I am also having some trouble building this with GCC 4.7.2 + OpenMPI 1.10.1. Lots of errors like:

  CXX      ../src/libSU2_AD_a-config_structure.o
In file included from ../../../Common/lib/../src/../include/./mpi_structure.hpp:58:0,
                 from ../../../Common/lib/../src/../include/config_structure.hpp:37,
                 from ../../../Common/lib/../src/config_structure.cpp:34:
../../../externals/medi/include/medi/codiMediPackTypes.hpp: In instantiation of ‘void CoDiPackToolBase<CoDiType, primalRestore, Impl>::addToolAction(medi::HandleBase*) [with CoDiType = codi::ActiveReal<codi::JacobiTape<codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> > > >; bool primalRestore = false; Impl = CoDiPackTool<codi::ActiveReal<codi::JacobiTape<codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> > > > >]’:
../../../externals/medi/include/medi/../../generated/medi/ampiFunctions.hpp:874:9:   required from ‘int medi::AMPI_Isend(const typename DATATYPE::Type*, int, DATATYPE*, int, int, MPI_Comm, medi::AMPI_Request*) [with DATATYPE = medi::MpiTypeDefault<CoDiPackTool<codi::ActiveReal<codi::JacobiTape<codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> > > > > >; typename DATATYPE::Type = codi::ActiveReal<codi::JacobiTape<codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> > > >; MPI_Comm = ompi_communicator_t*]’
../../../Common/lib/../src/../include/./mpi_structure.inl:216:85:   required from here
../../../externals/medi/include/medi/codiMediPackTypes.hpp:255:7: error: invalid conversion from ‘void (*)(void*, void*)’ to ‘codi::ExternalFunction::CallFunction {aka void (*)(void*)}’ [-fpermissive]
In file included from ../../../externals/codi/include/tapes/jacobiTape.hpp:195:0,
                 from ../../../externals/codi/include/codi.hpp:35,
                 from ../../../Common/lib/../src/../include/././datatypes/codi_reverse_structure.hpp:35,
                 from ../../../Common/lib/../src/../include/././datatype_structure.hpp:47,
                 from ../../../Common/lib/../src/../include/./mpi_structure.hpp:42,
                 from ../../../Common/lib/../src/../include/config_structure.hpp:37,
                 from ../../../Common/lib/../src/config_structure.cpp:34:
../../../externals/codi/include/tapes/modules/externalFunctionsModule.tpp:245:10: error:   initializing argument 1 of ‘void codi::JacobiTape<TapeTypes>::pushExternalFunctionHandle(codi::ExternalFunction::CallFunction, void*, codi::ExternalFunction::DeleteFunction) [with TapeTypes = codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> >; codi::ExternalFunction::CallFunction = void (*)(void*); codi::ExternalFunction::DeleteFunction = void (*)(void*)]’ [-fpermissive]
In file included from ../../../Common/lib/../src/../include/./mpi_structure.hpp:58:0,
                 from ../../../Common/lib/../src/../include/config_structure.hpp:37,
                 from ../../../Common/lib/../src/config_structure.cpp:34:
../../../externals/medi/include/medi/codiMediPackTypes.hpp:255:7: error: invalid conversion from ‘void (*)(void*, void*)’ to ‘codi::ExternalFunction::DeleteFunction {aka void (*)(void*)}’ [-fpermissive]
In file included from ../../../externals/codi/include/tapes/jacobiTape.hpp:195:0,
                 from ../../../externals/codi/include/codi.hpp:35,
                 from ../../../Common/lib/../src/../include/././datatypes/codi_reverse_structure.hpp:35,
                 from ../../../Common/lib/../src/../include/././datatype_structure.hpp:47,
                 from ../../../Common/lib/../src/../include/./mpi_structure.hpp:42,
                 from ../../../Common/lib/../src/../include/config_structure.hpp:37,
                 from ../../../Common/lib/../src/config_structure.cpp:34:
../../../externals/codi/include/tapes/modules/externalFunctionsModule.tpp:245:10: error:   initializing argument 3 of ‘void codi::JacobiTape<TapeTypes>::pushExternalFunctionHandle(codi::ExternalFunction::CallFunction, void*, codi::ExternalFunction::DeleteFunction) [with TapeTypes = codi::ChunkTapeTypes<double, codi::LinearIndexHandler<int> >; codi::ExternalFunction::CallFunction = void (*)(void*); codi::ExternalFunction::DeleteFunction = void (*)(void*)]’ [-fpermissive]
make[2]: *** [../src/libSU2_AD_a-config_structure.o] Error 1

talbring commented:
It looks like the codi submodule was not updated properly. Can you try to run the preconfigure command with the additional option "-u" ?

MaxSagebaum commented:
Maybe we could define the NO_OPTINAL_CONST based on the MPI Version.

https://stackoverflow.com/questions/17772906/why-does-mpi-send-accept-void-source

Gives a hint as to when the transition was made.

On Tue, 2017-11-21 at 19:53 +0000, Tim Albring wrote:

@talbring commented on this pull request.

________________________________

In .travis.yml<https://github.com/su2code/SU2/pull/473#discussion_r152384146>:

> @@ -20,7 +20,7 @@ branches:

 env:
     global:
-        CXXFLAGS="-O3 -Wall -Wextra -Wno-unused-parameter -Wno-empty-body -Wno-format-security"
+        CXXFLAGS="-O3 -Wall -Wextra -Wno-unused-parameter -Wno-empty-body -Wno-format-security -DMEDI_NO_OPTIONAL_CONST"


The problem is that depending on the actual MPI implementation (so for example MPICH, openmpi etc.) send buffer arguments might be declared as const or not. Unfortunately, there is no (easy) way to get that information automatically. It might be even different between versions of openmpi (on travis it is v1.6.5 and its necessary, locally I have v3.x, and its not)... so for now it is trial and error whether this is necessary or not. Maybe I can find that information somewhere, then I will provide it in the wiki.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/su2code/SU2/pull/473#discussion_r152384146>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMWwAeCRxzq9FO4-6pkSHTDnihbtByhvks5s4yoygaJpZM4QmNie>.

--

Max Sagebaum Chair for Scientific Computing, TU Kaiserslautern, Bldg/Geb 34, Paul-Ehrlich-Strasse, 67663 Kaiserslautern, Germany Phone: +49 (0)631 205 5638 Fax: +49 (0)631 205 3056 Email: max.sagebaum@scicomp.uni-kl.de URL: www.scicomp.uni-kl.de

talbring commented:
A short summary of the things I did in the additional commits:

- I added a consistent error handling that can be used throughout the code. For example if a marker is not specified the error looks like that: 
```
-------------------------------- Error ---------------------------------
In "short unsigned int CConfig::GetMarker_CfgFile_TagBound(std::__cxx11::string)": 
------------------------------------------------------------------------
The configuration file doesn't have any definition for marker airfoil
------------------------------------------------------------------------
Exiting now ...
```
or if a solution cannot be found:

```
-------------------------------- Error ---------------------------------
In "void CSolver::Read_SU2_Restart_Metadata(CGeometry*, CConfig*, bool, std::__cxx11::string)": 
------------------------------------------------------------------------
Unable to open SU2 restart file solution_flow.dat
------------------------------------------------------------------------
Exiting now ...
```
In case you want to throw an error, you can simply call the function `SU2_MPI::Error("Error Message", CURRENT_FUNCTION)` where `CURRENT_FUNCTION` is a predefined macro which expands to the name of the routine at compile time. I already replaced almost all error messages in the code.

- More MPI functions are included in the MPI Wrapper.

- The rank/size is now stored in most of the bigger classes and initialized in the constructor. That means there are no calls to MPI_Comm_rank / MPI_Comm_size anymore in every routine ... For convenience there is also a SU2_MPI::GetRank() / SU2_MPI::GetSize(), in case you need it outside of the main classes.

Let me know what you think. Not everything I did has to stay the way I implemented it :)

economon commented:
Looking ok for me now on ubuntu, but I am getting an error on OS X w/ Apple LLVM:

In file included from ../../../Common/lib/../src/mpi_structure.cpp:43:
In file included from ../../../externals/medi/src/medi/medi.cpp:29:
../../../externals/medi/src/medi/ampi/ampi.cpp:39:28: error: default initialization of an object of const type 'const medi::AMPI_IN_PLACE_IMPL' without a user-provided default constructor
  const AMPI_IN_PLACE_IMPL AMPI_IN_PLACE;
                           ^
                                        {}
296 warnings and 1 error generated.
make[2]: *** [../src/libSU2_AD_a-mpi_structure.o] Error 1

Also, lots of overloaded virtual warnings for multiple functions in the passive type header, such as:

../../../externals/medi/include/medi/ampi/../../../generated/medi/../../include/medi/ampi/typePassive.hpp:159:19: warning: 'medi::MpiTypePassive<medi::PairWithInt<long double> >::createModifiedTypeBuffer' hides overloaded virtual
      function [-Woverloaded-virtual]

Lastly, I really like the new error routine. Nice work. My only (picky) comments are on the formatting: could you please add a couple of empty lines of white space after the exit? The SU2 errors are getting lost with some other MPI clean up output and are not easy for the user to see immediately. One more personal preference: I would go without the line btwn the function and error message for clarity and, for consistency with the successful exit, remove the "Exiting now..." in favor of something like

------------------------- Error Exit (SU2_CFD) ------------------------

What do you think?

rsanfer commented on Common/src/grid_movement_structure.cpp:
```diff
@@ -9160,11 +9160,6 @@ void CElasticityMovement::SetVolume_Deformation_Elas(CGeometry *geometry, CConfi
 
   bool discrete_adjoint = config->GetDiscrete_Adjoint();
 
-  int rank = MASTER_NODE;
```
Hey Tim,
I've noticed that you remove this statements all over the place. However, "rank" and "rank=MASTER_NODE" is still widely used, how does the code access the variable?

rsanfer commented on SU2_CFD/src/numerics_direct_elasticity.cpp:
```diff
@@ -338,16 +338,13 @@ void CFEM_Elasticity::SetElement_Properties(CElement *element, CConfig *config)
 
 void CFEM_Elasticity::ReadDV(CConfig *config) {
 
+  int rank = SU2_MPI::GetRank();
```
Here it is declared...

rsanfer commented on SU2_CFD/src/solver_adjoint_elasticity.cpp:
```diff
@@ -91,10 +91,6 @@ CDiscAdjFEASolver::CDiscAdjFEASolver(CGeometry *geometry, CConfig *config, CSolv
   bool compressible = (config->GetKind_Regime() == COMPRESSIBLE);
   bool incompressible = (config->GetKind_Regime() == INCOMPRESSIBLE);
 
-  int rank = MASTER_NODE;
```
However in this file all the declarations are removed but rank == MASTER_NODE is used in lines 686, 708, 1154, 1162

talbring replied:
I made rank/size a member of most of the bigger classes and initialize it only in the constructor. Sometimes, for smaller classes, where it is just used one or two times, I still create a local variable.

talbring commented:
@economon the error will be fixed in the next commit.

About the error message: I added the additional dashed line in order to immediately see the error message. 

What about the following ?
```
...
------------------------ Iteration Preprocessing ------------------------
Zone 1: Euler/Navier-Stokes/RANS fluid iteration.

------------------------- Solver Preprocessing --------------------------

Error in "void CSolver::Read_SU2_Restart_Metadata(CGeometry*, CConfig*, bool, std::__cxx11::string)": 
-------------------------------------------------------------------------
Unable to open SU2 restart file solution_flow.dat
----------------------------- Exit Failure ------------------------------

```

Unfortunately I cannot get the kind of SU2 module in the interface ...

talbring commented:
What about the others ? If you have any doubts, comments or suggestions, please let me know. 

I am pretty sure that everybody will at least come across the error message sooner or later, so know is your chance to influence its style :)

economon commented:
The compile error is fixed on Apple LLVM for me now, thanks.

I am happy with the output format for the error now, with one small comment: I think that "Exit Failure" in the final line might be unclear to the user, in that, one might think that the code actually failed to exit correctly rather than exiting due to a runtime error. That's why I originally proposed "Error Exit," but I'm open to something else. No problem on not having the executable name (not critical at all).

talbring commented:
Tom, I see your point with the exit message. The current style is the following;

```
------------------------ Iteration Preprocessing ------------------------
Zone 1: Euler/Navier-Stokes/RANS fluid iteration.

------------------------- Solver Preprocessing --------------------------


Error in "void CSolver::Read_SU2_Restart_Metadata(CGeometry*, CConfig*, bool, std::__cxx11::string)": 
-------------------------------------------------------------------------
Unable to open restart file solution_flow.dat
------------------------------ Error Exit -------------------------------

```

economon commented:
Thanks, Tim. This PR looks good to me now.

Do you need any additional help finishing this one? Are the errors replaced everywhere and changes w.r.t. rank/size all sorted as you like?

@rsanfer: when you have time, could you please revisit your review?

All: I recommend that everyone tries to build this on your systems with your favorite compiler and flavor of MPI. Any comments/reviews from others are most welcome.

vdweide commented:
I built it with OpenMPI v2.1.1 in combination with gnu compilers, version 7.0.1, and all debug flags enabled. Apart from a lot of warnings in ParMETIS and CGNS, I also get the following warnings

../src/geometry_structure.cpp:16468:103: warning: variable ‘SemiSpan’ set but not used [-Wunused-but-set-variable]
   su2double Angle, MinAngle, MaxAngle, dAngle, *Area, *MaxThickness, *ToC, *Chord, *LERadius, *Twist, SemiSpan;

../src/output_structure.cpp:16249:46: warning: variable ‘Energy’ set but not used [-Wunused-but-set-variable]
   Velocity[3], Velocity2, MassFlow, Density, Energy, Area, AxiFactor = 1.0, SoundSpeed, Vn, Weight = 1.0;

../src/output_structure.cpp:16253:13: warning: unused variable ‘Gamma_Minus_One’ [-Wunused-variable]
   su2double Gamma_Minus_One   = Gamma - 1.0;

../src/output_structure.cpp:16256:18: warning: unused variable ‘nVar’ [-Wunused-variable]
   unsigned short nVar         = solver->GetnVar();

However, these are from an earlier merge. I already took care of them in feature_hom, so I suppose it can wait until we merge that branch into develop. So it is good to go as far as I am concerned.

rsanfer commented:
All good from my side. The tests are passing and I managed to compile (and cross check) the regression test cases for the discrete adjoint FEA and FSI both in my workstation at Imperial and even running in WSL (which I find amazing that SU2 runs in there). In this latest case, I used the openSUSE Leap 42.3 app for windows and the (pretty old) gcc 4.8.5 compilers shipped along. It worked.

From my side, this is ready to go. I'll remove the block straight away. 

Thanks, Tim and Max, for the efforts!

talbring commented:
I checked for last error messages and MPI functions. Everything seems to be fine. From my side it is ready to be merged :+1:
